pkg/azuredisk/controllerserver.go

```diff
@@ -1239,6 +1254,183 @@ func (d *Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReques
+			mc.ObserveOperationWithResult(isOperationSucceeded, consts.SourceResourceID, sourceVolumeID, consts.SnapshotName, snapshotNames[idx])
+		}
+	}()
+
+	klog.V(2).Infof("begin to create volume group snapshot(%s) in region(%s)", volumeGroupSnapshotName, d.cloud.Location)
+	volumeGroupReady := true
+	creationTime := timestamppb.New(time.Now())
+	snapshotList := []*csi.Snapshot{}
+	for idx := range snapshotNames {
+		createSnapshotReq := &csi.CreateSnapshotRequest{
+			Name:           snapshotNames[idx],
+			SourceVolumeId: sourceVolumeIDs[idx],
+			Parameters:     req.Parameters,
+			Secrets:        req.Secrets,
+		}
+		snapshotResp, err := d.CreateSnapshot(ctx, createSnapshotReq)
+		if err != nil {
+			return nil, err
+		}
+		if snapshotResp.Snapshot != nil {
+			if !snapshotResp.Snapshot.ReadyToUse {
+				volumeGroupReady = false
+			}
+			snapshotResp.Snapshot.GroupSnapshotId = volumeGroupSnapshotName
+			snapshotList = append(snapshotList, snapshotResp.Snapshot)
+			creationTime = snapshotResp.Snapshot.CreationTime
+		}
+	}
+	klog.V(2).Infof("create volume group snapshot(%s) in region(%s) successfully", volumeGroupSnapshotName, d.cloud.Location)
+	csiVolumeGroupSnapshot := &csi.VolumeGroupSnapshot{
+		GroupSnapshotId: volumeGroupSnapshotName,
+		Snapshots:       snapshotList,
+		CreationTime:    creationTime,
+		ReadyToUse:      volumeGroupReady,
+	}
+
+	createResp := &csi.CreateVolumeGroupSnapshotResponse{
+		GroupSnapshot: csiVolumeGroupSnapshot,
+	}
+	isOperationSucceeded = true
+	return createResp, nil
+}
+
+func (d *Driver) DeleteVolumeGroupSnapshot(ctx context.Context, req *csi.DeleteVolumeGroupSnapshotRequest) (*csi.DeleteVolumeGroupSnapshotResponse, error) {
+	groupSnapshotID := req.GroupSnapshotId
+	snapshotIDs := req.SnapshotIds
+	if len(groupSnapshotID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "GroupVolumeSnapshot ID must be provided")
+	}
+	if len(snapshotIDs) == 0 {
+		klog.V(2).Infof("GroupVolumeSnapshot is empty")
+		return &csi.DeleteVolumeGroupSnapshotResponse{}, nil
+	}
+
+	var err error
+	groupSnapshotName := groupSnapshotID
+	resourceGroup := d.cloud.ResourceGroup
+	snapshotNames := []string{}
+
+	for idx, snapshotID := range snapshotIDs {
+		if len(snapshotID) == 0 {
+			return nil, status.Error(codes.InvalidArgument, "Snapshot ID must be provided")
+		}
+		snapshotNames = append(snapshotNames, snapshotID)
+		if azureutils.IsARMResourceID(snapshotID) {
+			_, resourceGroup, snapshotNames[idx], err = azureutils.GetInfoFromURI(snapshotID)
+			if err != nil {
+				return nil, status.Errorf(codes.InvalidArgument, "%v", err)
+			}
+		}
+	}
+
+	mc := metrics.NewMetricContext(consts.AzureDiskCSIDriverName, "controller_delete_volume_group_snapshot", d.cloud.ResourceGroup, d.cloud.SubscriptionID, d.Name)
+	isOperationSucceeded := false
+	defer func() {
+		for _, snapshotID := range snapshotIDs {
+			mc.ObserveOperationWithResult(isOperationSucceeded, consts.SnapshotID, snapshotID)
```

Comment:
not necessary metrics here since we already have metrics in every DeleteSnapshot call

Reply:
done